pkg/cloudprovider/providers/aws/aws.go

```diff
@@ -2565,10 +2612,73 @@ func (c *Cloud) getTaggedSecurityGroups() (map[string]*ec2.SecurityGroup, error)
 	return m, nil
 }
 
+// Almost identical to updateInstanceSecurityGroupsForLoadBalancer, but it only adds ssg rules to
+// instances' inbound rules.
+func (s *AWSCloud) updateInstanceSharedSecurityGroups(ssgID string, allInstances []*ec2.Instance) error {
+	if s.cfg.Global.DisableSecurityGroupIngress || s.cfg.Global.EnableSharedSecurityGroupIngress == false {
+		return nil
+	}
+	taggedSecurityGroups, err := s.getTaggedSecurityGroups()
+	if err != nil {
+		return fmt.Errorf("error querying for tagged security groups: %v", err)
+	}
+
+	// Open the firewall from the load balancer to the instance
+	// We don't actually have a trivial way to know in advance which security group the instance is in
+	// (it is probably the minion security group, but we don't easily have that).
+	// However, we _do_ have the list of security groups on the instance records.
+
+	// Map containing the changes we want to make; true to add, false to remove
+	instanceSecurityGroupIds := map[string]bool{}
+
+	// Scan instances for groups we want open
+	for _, instance := range allInstances {
+		securityGroup, err := findSecurityGroupForInstance(instance, taggedSecurityGroups)
+		if err != nil {
+			return err
+		}
+
+		if securityGroup == nil {
+			glog.Warning("Ignoring instance without security group: ", orEmpty(instance.InstanceId))
+			continue
+		}
+		id := aws.StringValue(securityGroup.GroupId)
+		if id == "" {
+			glog.Warningf("found security group without id: %v", securityGroup)
+			continue
+		}
+
+		instanceSecurityGroupIds[id] = true
+	}
+
+	for instanceSecurityGroupId, _ := range instanceSecurityGroupIds {
+		sourceGroupId := &ec2.UserIdGroupPair{}
+		sourceGroupId.GroupId = &ssgID
+
+		// Instances should allow every traffic coming from the ssg resource.
+		allProtocols := "-1"
+
+		permission := &ec2.IpPermission{}
+		permission.IpProtocol = &allProtocols
+		permission.UserIdGroupPairs = []*ec2.UserIdGroupPair{sourceGroupId}
+
+		permissions := []*ec2.IpPermission{permission}
+
+		changed, err := s.addSecurityGroupIngress(instanceSecurityGroupId, permissions)
+		if err != nil {
+			return err
+		}
+		if !changed {
+			glog.Warning("shared security group id already in instance's security group!")
+		}
+	}
+	return nil
+}
+
 // Open security group ingress rules on the instances so that the load balancer can talk to them
 // Will also remove any security groups ingress rules for the load balancer that are _not_ needed for allInstances
 func (c *Cloud) updateInstanceSecurityGroupsForLoadBalancer(lb *elb.LoadBalancerDescription, allInstances []*ec2.Instance) error {
-	if c.cfg.Global.DisableSecurityGroupIngress {
+	if c.cfg.Global.DisableSecurityGroupIngress || c.cfg.Global.EnableSharedSecurityGroupIngress {
```

Comment:
I feel we have the switching logic spread across several functions now, and it would be clearer if that was not the case